### PR TITLE
ReactOS 1st-stage GUI Setup: Add the update-registry code - Part 3

### PR DESCRIPTION
**Pull request by HBelusca**
## Purpose

Add the update-registry code.
This code has been tested on Windows and works as expected.

## Proposed changes

- The code adds a call to `UpdateRegistry()` function (which does all the job), and the corresponding UI callback `RegistryStatus()`.
- Translation (work in progress) is added;
- Extra helpers for simplifying LoadString+printf+setting window text have been added.

## TODO

- [x] Rebase this PR once the other PRs #7159 and #7162 are merged in master.
- [x] Finish the translations.

### CODE REVIEW

File: base/setup/lib/setuplib.c [hunk 1]
```diff
@@ -477,7 +477,7 @@ GetSourcePaths(
 
     /* Check whether the resolved \SystemRoot is a prefix of the image file path */
     // FIXME: commented out to allow installation from USB
-    // if (RtlPrefixUnicodeString(&SystemRootPath, InstallSourcePath, TRUE))
+    /**/if (RtlPrefixUnicodeString(&SystemRootPath, InstallSourcePath, TRUE))/**/
```
Comment (JoachimHenze) [inline base/setup/lib/setuplib.c:480]: I would encourage you to undo that change.
```suggestion
    // if (RtlPrefixUnicodeString(&SystemRootPath, InstallSourcePath, TRUE))
```

File: base/setup/reactos/lang/en-US.rc [hunk 2]
```diff
@@ -173,3 +173,30 @@ BEGIN
...
     IDS_PARTITION_STATUS "Status"
 END
+
+/* Progress Strings */
+
+STRINGTABLE
+BEGIN
+    ///// IDS_PREPARE_PARTITIONS "Preparing partitions..."
+    ///// IDS_PREPARE_FILES "Preparing the list of files to be copied, please wait..."
+    ///// IDS_COPYING_FILES "Copying the files..."
```
Comment (JoachimHenze) [inline base/setup/reactos/lang/en-US.rc:183]:
```suggestion
    //IDS_PREPARE_PARTITIONS "Preparing partitions..."
    //IDS_PREPARE_FILES "Preparing the list of files to be copied, please wait..."
    //IDS_COPYING_FILES "Copying the files..."
```

File: base/setup/reactos/reactos.c [hunk 3]
```diff
@@ -92,17 +95,191 @@ CreateTitleFont(VOID)
...
+
+    /* We need to retrieve the current module's instance handle if either
+     * the title or the format message is specified by a resource ID */
+    if ((pszTitle && IS_INTRESOURCE(pszTitle)) || IS_INTRESOURCE(pszFormatMessage))
+        hInstance = GetModuleHandleW(NULL); // SetupData.hInstance;
+
+    /* Retrieve the format message string if this is a resource */
+    if (pszFormatMessage && IS_INTRESOURCE(pszFormatMessage)) do
+    {
+        // LoadAllocStringW()
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:121]:
```suggestion
```

File: base/setup/reactos/reactos.c [hunk 4]
```diff
@@ -92,17 +95,191 @@ CreateTitleFont(VOID)
...
+        PCWSTR pStr;
+
+        /* Try to load the string from the resource */
+        MsgLen = LoadStringW(hInstance, PtrToUlong(pszFormatMessage), (LPWSTR)&pStr, 0);
+        if (MsgLen == 0)
+        {
+            /* No resource string was found, return NULL */
+            Format = NULL;
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:129]: Consider moving that assignment upwards before the line `if (pszFormatMessage && IS_INTRESOURCE(pszFormatMessage)) do`
That would guarantee Format to be NULL even in case that somebody else will add some other break into this do-while(0)-block.

File: base/setup/reactos/reactos.c [hunk 5]
```diff
@@ -92,17 +95,191 @@ CreateTitleFont(VOID)
...
+         * The string is built to be NULL-terminated.
+         */
+        MsgLen = _vscwprintf(Format, args);
+        if (MsgLen >= _countof(StaticBuffer))
+        {
+            Buffer = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, (MsgLen + 1) * sizeof(WCHAR));
+            if (!Buffer)
+            {
+                /* Allocation failed, use the original format string verbatim */
+                // MsgLen = wcslen(Format);
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:163]:
```suggestion
```
strip it or implement it?

File: base/setup/reactos/reactos.c [hunk 6]
```diff
@@ -92,17 +95,191 @@ CreateTitleFont(VOID)
...
+                Buffer = Format;
+            }
+        }
+        if (Buffer != Format)
+        {
+            /* Do the printf as we use the caller's format string */
+            // _vsnwprintf(Buffer, MsgLen, Format, args);
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:170]:
```suggestion
```
strip it or implement it?

File: base/setup/reactos/reactos.c [hunk 7]
```diff
@@ -92,17 +95,191 @@ CreateTitleFont(VOID)
...
+    mb.cbSize = sizeof(mb);
+    mb.hwndOwner = hWnd;
+    mb.hInstance = hInstance;
+    mb.lpszText = Buffer;
+    mb.lpszCaption = pszTitle;
+    mb.dwStyle = uType;
+    mb.dwLanguageId = MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT);
+
+    iRes = MessageBoxIndirectW(&mb);
+    }
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:192]:
```suggestion
    {
        MSGBOXPARAMSW mb = {0};
        mb.cbSize = sizeof(mb);
        mb.hwndOwner = hWnd;
        mb.hInstance = hInstance;
        mb.lpszText = Buffer;
        mb.lpszCaption = pszTitle;
        mb.dwStyle = uType;
        mb.dwLanguageId = MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT);

        iRes = MessageBoxIndirectW(&mb);
    }
```

File: base/setup/reactos/reactos.c [hunk 8]
```diff
@@ -1079,14 +1254,48 @@ FileCopyCallback(PVOID Context,
...
+        IDS_REG_ADDKBLAYOUTS,           /* KeybLayouts */
+        IDS_REG_KEYBOARDSETTINGSUPDATE, /* KeybSettingsUpdate */
+        IDS_REG_CODEPAGEINFOUPDATE,     /* CodePageInfoUpdate */
+    };
+
+    if (RegStatus < _countof(StringIDs))
+    {
+        va_list args;
+        va_start(args, RegStatus);
+        // GetDlgItem(UiContext.hwndDlg, IDC_ITEM)
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:1286]:
```suggestion
```
a preparation for the future? Or forgotten to strip that?

File: base/setup/reactos/reactos.c [hunk 9]
```diff
@@ -1079,14 +1254,48 @@ FileCopyCallback(PVOID Context,
...
+        SetWindowResPrintfVW(UiContext.hWndItem, SetupData.hInstance, StringIDs[RegStatus], args);
+        va_end(args);
+    }
+    else
+    {
+        // GetDlgItem(UiContext.hwndDlg, IDC_ITEM)
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.c:1292]:
```suggestion
```
a preparation for the future? Or forgotten to strip that?

File: base/setup/reactos/reactos.h [hunk 10]
```diff
@@ -108,6 +111,15 @@ typedef struct _NT_WIN32_PATH_MAPPING_LIST
...
 
+#if 0
+typedef struct _KBLAYOUT
+{
+    TCHAR LayoutId[9];
+    TCHAR LayoutName[128];
+    TCHAR DllName[128];
+} KBLAYOUT, *PKBLAYOUT;
+#endif
+
```
Comment (JoachimHenze) [inline base/setup/reactos/reactos.h:122]: Why did you move that typedef to the bottom? Can you move it back up to aovid the useless diff or wouldn't that compile anymore?

File: base/setup/reactos/resource.h [hunk 11]
```diff
@@ -100,3 +100,22 @@
 #define IDS_PARTITION_TYPE   5201
 #define IDS_PARTITION_SIZE   5202
 #define IDS_PARTITION_STATUS 5203
+
+
+/* Progress Strings */
+/////// #define IDS_PREPARE_PARTITIONS      5308
+/////// #define IDS_PREPARE_FILES           5309
+/////// #define IDS_COPYING_FILES           5310
```
Comment (JoachimHenze) [inline base/setup/reactos/resource.h:108]:
```suggestion
//#define IDS_PREPARE_PARTITIONS      5308
//#define IDS_PREPARE_FILES           5309
//#define IDS_COPYING_FILES           5310
```

File: base/setup/lib/setuplib.c (hunk 1 above)
Comment (HBelusca) [reply base/setup/lib/setuplib.c:468]: This is from the 1st temporary commit, will go away in the final PR version.

File: base/setup/reactos/lang/en-US.rc (hunk 2 above)
Comment (HBelusca) [reply base/setup/reactos/lang/en-US.rc:195]: Yep, these were markers for me to indicate where the stuff from PR #7159 will land to. Will disappear in the final PR version.

**Review by GeoB99 — APPROVED** [review]

File: base/setup/reactos/reactos.c (hunk 9 above)
Comment (HBelusca) [reply base/setup/reactos/reactos.c]: forgot to strip, i've removed it.

File: base/setup/reactos/reactos.c (hunk 8 above)
Comment (HBelusca) [reply base/setup/reactos/reactos.c]: forgot to strip, i've removed it.

**Comment by HBelusca** [comment]
All TODOs and review comments addressed, good to go.
